### ISSUE TITLE
Proposal: Windows-friendly build scripts

### ISSUE THREAD

**Issue by ajmchambers** [cyclejs/cyclejs#481]
Quick summary (as I have the tendency of writing walls of text): I'm having troubles running this repo's setup and tests in Windows. I also have a Mac that I can test so I'm willing to make pull requests to fix any issues that come up.

I'm creating this issue to show an example of what some the issues are and what the changes would look like... and also because I read the contributing guide which told me to:

> 1. Open a GitHub issue before doing significant amount of work.

----------------------------------------

## `.scripts/npm-x-for-all-packages.sh`
 
I cloned this repo and ran `npm run setup` which ended with this result:

```bash
> @ internal-install-all C:\Development\forks\cyclejs
> .scripts/npm-x-for-all-packages.sh install

'.scripts' is not recognized as an internal or external command,
operable program or batch file
```

I'm using bash emulation on windows (using [cmder](http://cmder.net/)) so I attempted to run the script with `sh .scripts/npm-x-for-all-packages.sh install`.

This actually ran, but failed in a different way (and even if it had of worked it still wouldn't have worked on regular windows command prompt).

This is what I got:

![sherror](https://cloud.githubusercontent.com/assets/5426717/20733750/0ee3fe38-b68d-11e6-83ef-f37cec9f13d7.png)

So this led me to create a node equivalent version of the `npm-x-for-all-packages.sh`, here's a link -> [`npm-x-for-all-packages.js`](https://github.com/ajmchambers/cyclejs/blob/cycle-base-scripts-windows/.scripts/npm-x-for-all-packages.js)

Then I modified the `package.json`:

```json
"internal-install-all": "node .scripts/npm-x-for-all-packages.js install",
```

This works for me in windows command prompt, git-bash and also mac terminal.

So with that out of the way I attempted to: 
```bash
node .scripts/npm-x-for-all-packages.js run test
``` 

This works fine on mac, but fails on windows.

Which led me to...

## `cyclejs/base/package.json`

I decided to try and get all the scripts in `base` working and ran into a couple of issues:

1. `"prelib"` script

 ```json
 "prelib": "rm -rf lib/ && mkdir -p lib/",
 ```

This can be fixed using [rimraf](https://www.npmjs.com/package/rimraf) and [mkdirp](https://www.npmjs.com/package/mkdirp), so I added these packages to the root `package.json`'s devDependencies so that they can be used across the other packages.

```json
 "prelib": "../node_modules/.bin/rimraf lib/ && ../node_modules/.bin/mkdirp lib/", 
```

Which led me to another windows issue: 

```bash
> @cycle/base@4.1.1 prelib C:\Development\forks\cyclejs\base
> ../node_modules/.bin/rimraf lib/ && ../node_modules/.bin/mkdirp lib/

'..' is not recognized as an internal or external command,
```

This is an issue with npm scripts and windows: 
https://github.com/npm/npm/issues/13789

However if you change the script to this:

```json
  "prelib": "node ../node_modules/rimraf/bin.js lib/ && node ../node_modules/mkdirp/bin/cmd.js lib/", 
```

It will work correctly on command prompt, gitbash, and mac terminal.

So I went and updated the rest of the scripts in the base [`package.json`](https://github.com/ajmchambers/cyclejs/blob/cycle-base-scripts-windows/base/package.json).

Which looks like this:

```json
  "scripts": {
    "lint": "node ../node_modules/tslint/bin/tslint -c ../tslint.json src/**/*.ts",
    "prelib": "node ../node_modules/rimraf/bin.js lib/ && node ../node_modules/mkdirp/bin/cmd.js lib/",
    "lib": "node ../node_modules/typescript/bin/tsc",
    "premocha": "npm run lib",
    "mocha": "node ../node_modules/mocha/bin/mocha --compilers js:babel-register test",
    "test": "npm run lint && npm run mocha",
    "test-ci": "npm run test",
    "prebrowserify": "npm run lib",
    "browserify": "node ../node_modules/browserify/bin/cmd.js lib/index.js -t babelify --standalone CycleBase -o dist/cycle-base.js",
    "uglify": "node ../node_modules/uglify-js/bin/uglifyjs dist/cycle-base.js -o dist/cycle-base.min.js",
    "predist": "node ../node_modules/rimraf/bin.js dist/ && node ../node_modules/mkdirp/bin/cmd.js dist/",
    "dist": "npm run browserify && npm run uglify",
    "readme": ":",
    "changelog": "node ../.scripts/update-changelogs.js ${PWD##*/}",
    "preversion": "npm test",
    "version": "npm run dist && npm run readme && npm run changelog",
    "postversion": "git add -A && git commit -m \"release(${PWD##*/}): v$(cat package.json | ../node_modules/.bin/jase version)\" && git push origin master && npm publish",
    "release-patch": "false",
    "release-minor": "npm --no-git-tag-version version minor",
    "release-major": "npm --no-git-tag-version version major"
  },
```

---------------------------------------

So basically if these kind of changes seem alright to make, I'd be happy to go through all the packages one-by-one and get them all working on windows.

Note that I'm still pretty new to github etc so I'm not sure what the best way would be to orchestrate all the changes: many small pull requests or fewer larger ones? so any advice there would be appreciated.

**Comment by whitecolor** [comment]
This problem had been [already raised](https://github.com/cyclejs/cyclejs/issues/384). 

But the [problem is in NPM](https://github.com/npm/npm/issues/13789). I just have made a [PR to Yarn](https://github.com/yarnpkg/yarn/pull/2082) to fix this.

As workaround you may fix your fork of NPM/Yarn or I even just fix it manually in npm installation (as long as I don't install npm often it is ok)

For yarn I hope it will be fixed, maybe then in NPM. Someday, somehow =)

**Comment by staltz** [comment]
Hi, thank you for taking your time to send us this issue. I'm sorry but our bash scripts are serving us very well, and it's not worth the time trying to support Windows. See this issue https://github.com/cyclejs/cyclejs/issues/384

I recommend installing Ubuntu in a virtual machine. It's free.

**Comment by ajmchambers** [comment]
Thats fair enough thanks, I understand not wanting to mess with around with the bash scripts - especially since you're not likely to be building or managing the overall repo on a windows machine anyway.

In that case I agree that the `internal-install-all` stuff and so on is not so important and its not a big deal to just manually `npm install` in the relevant folders that you care about.

However build scripts out of the way, it would still be beneficial to be able to run and test the packages locally on windows.

Assuming we just wait for the relative path bugs that @whitecolor mentioned to be fixed in npm / yarn, that leaves a couple of simple changes that could be made to get the tests working on windows:

- Replacing `mkdir -p` with `mkdirp` and `rm -rf` with `rimraf` in scripts like `"prelib"` would fix that issue.
- Adding some notes to `CONTRIBUTING.md` to make a note of the issues with windows eg:
  - not officially supported but if you want to run the tests then here is how you might go about it...
    - npm install manually
    - run at least version of npm / yarn (that has hopefully been fixed)
    - alternatively run in a VM, bash for windows etc....

**OR alternatively...**

Could take it a step further (and I mentioned it in my previous message), it is possible to modify `package.json` scripts to work around the npm/yarn relative path bug in the meantime by running them with node directly.

That would get the tests etc working in the meantime, then when npm / yarn is fixed the change can be reverted and simply leave a note for contributors saying: 'run at least this version of npm/yarn for it to work on windows'.

```json
    "lint": "node ../node_modules/tslint/bin/tslint -c ../tslint.json src/**/*.ts",
    "prelib": "node ../node_modules/rimraf/bin.js lib/ && node ../node_modules/mkdirp/bin/cmd.js lib/",
    "lib": "node ../node_modules/typescript/bin/tsc",
    "premocha": "npm run lib",
    "mocha": "node ../node_modules/mocha/bin/mocha --compilers js:babel-register test",
    "test": "npm run lint && npm run mocha",
    "test-ci": "npm run test",
    "prebrowserify": "npm run lib",
    "browserify": "node ../node_modules/browserify/bin/cmd.js lib/index.js -t babelify --standalone CycleBase -o dist/cycle-base.js",
    "uglify": "node ../node_modules/uglify-js/bin/uglifyjs dist/cycle-base.js -o dist/cycle-base.min.js",
    "predist": "node ../node_modules/rimraf/bin.js dist/ && node ../node_modules/mkdirp/bin/cmd.js dist/",
    "dist": "npm run browserify && npm run uglify",
```

I'd be happy to make submit pull requests for changes myself but again it's up to you whether you view this as a worthwhile effort, so if you are still not keen just let me know and I'll leave it be.

Cheers
